src/VisualStudio/CSharp/Impl/LanguageService/CSharpCodeCleanupFixer.cs

```diff
@@ -272,43 +279,56 @@ public override Task<bool> FixAsync(ICodeCleanUpScope scope, ICodeCleanUpExecuti
 
         private Task<bool> FixHierarchyContentAsync(IVsHierarchyCodeCleanupScope hierarchyContent, ICodeCleanUpExecutionContext context, CancellationToken cancellationToken)
         {
-            // TODO: this one will be implemented later
-            // https://github.com/dotnet/roslyn/issues/30165
             var hierarchy = hierarchyContent.Hierarchy;
             if (hierarchy == null)
             {
-                // solution
-                return Task.FromResult(false);
+                return FixSolutionAsync(_workspace.CurrentSolution, context, cancellationToken);
             }
 
             var itemId = hierarchyContent.ItemId;
             if (itemId == (uint)VSConstants.VSITEMID.Root)
             {
-                // Project
+                var hierarchyToProjectMap = _workspace.Services.GetRequiredService<IHierarchyItemToProjectIdMap>();
+                var projectHierarchyItem = _vsHierarchyItemManager.GetHierarchyItem(hierarchyContent.Hierarchy, itemId);
+                if (!hierarchyToProjectMap.TryGetProjectId(projectHierarchyItem, targetFrameworkMoniker: null, out var projectId))
+                {
+                    return Task.FromResult(false);
+                }
+
+                var project = _workspace.CurrentSolution.GetProject(projectId);
+                if (project is null)
+                {
+                    return Task.FromResult(false);
+                }
+
+                return FixProjectAsync(project, context, cancellationToken);
             }
             else if (hierarchy.GetCanonicalName(itemId, out var path) == 0)
             {
                 var attr = File.GetAttributes(path);
                 if (attr.HasFlag(FileAttributes.Directory))
                 {
                     // directory
+                    // TODO: this one will be implemented later
+                    // https://github.com/dotnet/roslyn/issues/30165
                 }
                 else
                 {
                     // document
+                    // TODO: this one will be implemented later
+                    // https://github.com/dotnet/roslyn/issues/30165
                 }
             }
 
             return Task.FromResult(false);
         }
 
-        private async Task<bool> FixTextBufferAsync(TextBufferCodeCleanUpScope textBufferScope, ICodeCleanUpExecutionContext context, CancellationToken cancellationToken)
+        private async Task<bool> FixSolutionAsync(Solution solution, ICodeCleanUpExecutionContext context, CancellationToken cancellationToken)
         {
             using (var cancellationTokenSource = CancellationTokenSource.CreateLinkedTokenSource(context.OperationContext.UserCancellationToken, cancellationToken))
```

Comment:
just curious, why there are 2 cancellation token? I get what user cancellation token is for, who owns the cancellationToken? what can raise that token?

Reply:
@olegtk ?

Reply:
indeed. @crmann1 ?